third_party/move/move-compiler-v2/src/function_checker.rs

```diff
@@ -89,156 +93,175 @@ pub fn check_access_and_use(env: &mut GlobalEnv) {
-                            functions_with_callers.insert(callee);
-                        } else {
-                            functions_with_inaccessible_callers.insert(callee);
+                        // Only record and warn about unused functions before inlining:
+                        if before_inlining {
+                            // Record called functions for Unused check below.
+                            if callee_is_accessible {
+                                functions_with_callers.insert(*callee);
+                            } else {
+                                functions_with_inaccessible_callers.insert(*callee);
+                            }
                         }
                     }
                 }
             }
         }
     }
 
-    // Check for Unused functions: private (or friendless public(friend)) funs with no callers.
-    let options = env
-        .get_extension::<Options>()
-        .expect("Options is available");
-    if options.warn_unused {
-        for callee in private_funcs {
-            if !functions_with_callers.contains(&callee) {
-                // We saw no uses of private/friendless function `callee`.
-                let callee_env = env.get_function(callee);
-                let callee_loc = callee_env.get_loc();
-                let callee_is_script = callee_env.module_env.get_name().is_script();
-
-                // Entry functions in a script don't need any uses.
-                // Check others which are private.
-                if !callee_is_script {
-                    let is_private = matches!(callee_env.visibility(), Visibility::Private);
-                    if functions_with_inaccessible_callers.contains(&callee) {
-                        let msg = format!(
-                            "Function `{}` may be unused: it has callers, but none with access.",
-                            callee_env.get_full_name_with_address(),
-                        );
-                        env.diag(Severity::Warning, &callee_loc, &msg);
-                    } else {
-                        let msg = format!(
-                            "Function `{}` is unused: it has no current callers and {}.",
-                            callee_env.get_full_name_with_address(),
-                            if is_private {
-                                "is private to its module"
-                            } else {
-                                "is `public(friend)` but its module has no friends"
-                            }
-                        );
-                        env.diag(Severity::Warning, &callee_loc, &msg);
+    if before_inlining {
+        // Check for Unused functions: private (or friendless public(friend)) funs with no callers.
+        let options = env
+            .get_extension::<Options>()
+            .expect("Options is available");
+        if options.warn_unused {
+            for callee in private_funcs {
```

Comment:
Are we missing out on warning a `public (friend)` function `foo` that is in a module with friends, but none of those friends call `foo`? Do we want in such cases as well?

Reply:
Good point.